hamcrest/src/main/java/org/hamcrest/collection/IsMapContaining.java

```diff
@@ -72,9 +72,32 @@ public void describeTo(Description description) {
      *     the value that, in combination with the key, must be describe at least one entry
      */
     public static <K,V> Matcher<Map<? extends K,? extends V>> hasEntry(K key, V value) {
-        return new IsMapContaining<>(equalTo(key), equalTo(value));
+        return new IsMapContainingEntry<>(key, value);
     }
-    
+
+    /**
+     * Provides a type-safe optimization over the O(n) linear search in {@link IsMapContaining#matchesSafely(Map)},
+     * by leveraging the speed of the map's own {@link Map#containsKey(Object)} check.
+     * <p>
+     * It preserves the same descriptors.
+     */
+    private static class IsMapContainingEntry<K, V> extends IsMapContaining<K, V>
+    {
+        private final K key;
+
+        public IsMapContainingEntry(K key, V value)
+        {
+            super(equalTo(key), equalTo(value));
+            this.key = key;
+        }
+
+        @Override
+        public boolean matchesSafely(Map<? extends K, ? extends V> map)
+        {
+            return map.containsKey(key) && super.valueMatcher.matches(map.get(key));
```

Comment:
There are maps that don't allow null keys and (stupidly) throw an exception when you try to check for its existence. You need to account for them here.
```suggestion
            try {
              return map.containsKey(key) && super.valueMatcher.matches(map.get(key));
            catch (NullPointerException ex) {
              return false;
            }
```